RFCs/FS-1063-support-letbang-andbang-for-applicative-functors.md

```diff
@@ -0,0 +1,769 @@
+# F# RFC FS-1063 - Support let! .. and... for applicative functors
+
+The design suggestion [Support let! .. and... for applicative functors](https://github.com/fsharp/fslang-suggestions/issues/579) has been marked "approved in principle".
+This RFC covers the detailed proposal for this suggestion.
+
+* [x] [Approved in principle](https://github.com/fsharp/fslang-suggestions/issues/579#event-1345361104) & [prioritised](https://github.com/fsharp/fslang-suggestions/issues/579#event-1501977428)
+* [x] [Suggestion](https://github.com/fsharp/fslang-suggestions/issues/579)
+* [x] [Discussion](https://github.com/fsharp/fslang-design/issues/335)
+* [ ] Implementation: [In progress](https://github.com/Microsoft/visualfsharp/pull/5696)
+
+# Summary
+[summary]: #summary
+
+Extend computation expressions to support applicative functors via a new `let! ... and! ... return ...` syntax.
+
+# Motivation
+[motivation]: #motivation
+
+Applicative functors (or just "applicatives", for short) have been growing in popularity as a way to build applications and model certain domains over the last decade or so, since McBride and Paterson published [Applicative Programming with Effects](http://www.staff.city.ac.uk/~ross/papers/Applicative.html). Applicatives are now reaching a level of popularity within the community that supporting them with a convenient and readable syntax, as we do for monads, makes sense.
+
+With applicative computation expressions, we can write more computations with this convenient syntax than before (there are more contexts which meet the requirements for applicative computation expressions than the existing monadic ones), and we can write more efficient computations (the requirements of applicatives rule out needing to support some potentially expensive operations).
+
+For example, [Pauan points out](https://github.com/fsharp/fslang-suggestions/issues/579#issuecomment-310799948) that we can write a convenient and readable computation expression for `Observable`s that [avoids unnecessary resubscriptions](https://github.com/fsharp/fslang-suggestions/issues/579#issuecomment-310854419) and syntactically scales nicely with the number of arguments:
+
+```fsharp
+// Outputs a + b + c, which is recomputed every time foo or bar outputs a new value,
+// avoiding any unnecessary resubscriptions
+observable {
+  let! a = foo
+  and! b = bar
+  and! c = baz
+  return a + b + c
+}
+```
+
```

Comment:
Can you add a sentence or two that explains what you would need to do without such a syntax? Just has to be high-level, but that helps cement the motivation.

Reply:
Actually I think this is fine without that.